jax/_src/extend/ffi.py

```diff
@@ -139,31 +147,36 @@ def _lowering(
     if "result_types" not in kwargs:
       kwargs["result_types"] = [mlir.aval_to_ir_type(aval) for aval in ctx.avals_out]
     if operand_layouts is None:
-      kwargs["operand_layouts"] = _default_layouts(aval.shape for aval in ctx.avals_in)  # pytype: disable=attribute-error
+      kwargs["operand_layouts"] = _default_layouts(ctx.avals_in)  # pytype: disable=attribute-error
     else:
       kwargs["operand_layouts"] = operand_layouts
     if result_layouts is None:
-      kwargs["result_layouts"] = _default_layouts(aval.shape for aval in ctx.avals_out)
+      kwargs["result_layouts"] = _default_layouts(ctx.avals_out)
     else:
       kwargs["result_layouts"] = result_layouts
     if "result_shapes" not in kwargs and not all(
-        core.is_constant_shape(aval.shape) for aval in ctx.avals_out):
+        core.is_constant_shape(_aval_shape(aval)) for aval in ctx.avals_out):
       kwargs["result_shapes"] = [
-          mlir.shape_tensor(mlir.eval_dynamic_shape_as_ivals(ctx, aval.shape))
+          mlir.shape_tensor(mlir.eval_dynamic_shape_as_ivals(ctx, _aval_shape(aval)))
           for aval in ctx.avals_out]
 
     return mlir.custom_call(call_target_name, operands=operands, **kwargs).results  # type: ignore
 
   return _lowering
 
 
-def _default_layouts(shapes: Iterable[Sequence[DimSize]]) -> list[list[DimSize]]:
-  return [list(reversed(range(len(shape)))) for shape in shapes]
+def _result_avals(results: Sequence[core.Value]) -> tuple[core.AbstractValue]:
+  avals = []
+  for result in results:
+    if hasattr(result, "dtype"):
+      _check_shape_dtype(result)
+    avals.append(core.raise_to_shaped(core.get_aval(result)))
+  return tuple(avals)
 
 
 def ffi_call(
     target_name: str,
-    result_shape_dtypes: DuckTypedArray | Sequence[DuckTypedArray],
+    results: core.Value | Sequence[core.Value],
```

Comment:
I don't think this type hint is very useful. Maybe `DuckTypedArray | AbstractToken`?

Also, I'd probably still prefer to call this `result_shape_dtypes` or similar to make it clear that this doesn't need to actually include arrays. I think that tokens are an advanced enough feature that it would be fine to keep the name and update the docstring. WDYT?

Reply:
Yeah, I unfortunately didn't see a particularly uniform solution short of passing `jax.core.Value`s here, but focussing on the common case here sounds fine to me. To be clear, you are suggesting to go with `result_shape_dtypes: DuckTypedArray | AbstractToken | Sequence[DuckTypedArray | AbstractToken]`?

Is there a hard and fast rule on what's considered public? (Both `Value` and `AbstractToken/abstract_token` are exported by `jax.core`.)